ddl/job_table.go

```diff
@@ -194,28 +165,51 @@ func hasSysDB(job *model.Job) bool {
 	return false
 }
 
-func (d *ddl) handleUpgradingState(se *sess.Session, job *model.Job) error {
-	if !d.stateSyncer.IsUpgradingState() {
-		if !job.IsPausedBySystem() || hasSysDB(job) {
-			return nil
+func (d *ddl) processJobDuringUpgrade(sess *sess.Session, job *model.Job) (isRunnable bool, err error) {
+	if d.stateSyncer.IsUpgradingState() {
+		if hasSysDB(job) {
+			return true, nil
 		}
-		_, err := ResumeJobsBySystem(se.Session(), []int64{job.ID})
-		if err != nil {
-			logutil.BgLogger().Warn("[ddl] resume user DDL by system failed", zap.Stringer("job", job), zap.Error(err))
-			return err
+		var errs []error
+		// During binary upgrade, pause all running DDL jobs
+		errs, err = PauseJobsBySystem(sess.Session(), []int64{job.ID})
+		if len(errs) > 0 {
+			if dbterror.ErrPausedDDLJob.Equal(errs[0]) {
+				// In case of 'Pausing', we need to deliver the job to DDL
+				// worker to pause the background reorganization workers.
+				return true, nil
+			}
+			if dbterror.ErrCannotPauseDDLJob.Equal(errs[0]) {
+				// The state could be 'cancelling' or 'rollback', as well as
+				// some corner state
+				return false, nil
+			}
```

Comment:
I think will only do `job.IsPausing || hasSysDB(job)` like https://github.com/pingcap/tidb/pull/43907/files#diff-91c692cc214077acc788fcaf2f92fb40d0a6c4aef3e34e3b6ed532677dac9f13L210.
Then we needn't distinguish the error msg, and needn't do line197.

Reply:
It is not a good idea to take care of the job's state out of 'PauseJobsBySystem'. 

As the function 'PauseJobsBySystem' will do the check and pause the job accordingly, we should only check the result of it but not moving some of its logical.

As the logic you mentioned:
1. If hasSysDB(), it will continue the running
2. If ErrPausedDDLJob returned from 'PauseJobsBySystem', then the job could be 'paused' or 'pausing'. And no matter what the state is, the ddl worker will handle it accordingly. Then we don't need to do the check here.

It is about cohesion.

Reply:
I means:
1. If `hasSysDB()`, it will continue the running
2. If `job.IsPausing`, it will continue the running(pausing to paused)
3. Do `PauseJobsBySystem`, if it encouters return `false, err`

Reply:
That is the same with current code.

Reply:
But we don't need to determine the error type. That can remove one `if...`

Reply:
Error handling is necessary when we are calling a function.

Once it returned the 'ErrPausedDDLJob', which means the job has already been 'paused' or 'pausing', we could just let it go running; otherwise, we should not run it.

We should implement the logic by calling a function and handle its return value but not moving some logic from it.